### ISSUE TITLE
Clean up command processing

### ISSUE THREAD

**Issue by seventh-chord** [simon-whitehead/hakka#20]
Currently command processing consists of one giant string containing help information for all the commands and one if-else statement to process them.  

I would suggest having a `Command` struct, with fields for `name`, `help_text` and a function to be executed when the command is run. We can then have a list/hashmap of commands which is searched when a command is executed.

**Comment by simon-whitehead** [comment]
I was thinking about using something like [clap.rs](https://clap.rs) instead, since its very fully featured and does long/short form commands, help text, etc.

Specifically, clap.rs has an [API for handling whatever input you throw at it](https://docs.rs/clap/2.19.3/clap/struct.App.html#method.get_matches_from) and not just from stdin. So we could tie that into the `Console` somehow.

What are your thoughts on that? I've not tested this at all, its just an idea. I am open to writing it all by hand as well but I figured its a good idea to investigate clap.rs anyway.

**Comment by simon-whitehead** [comment]
As a quick example (I spent about 20 minutes on this before rushing out to dinner, so its not bullet proof and may not be the best usage of their API?), here is a playground example using clap: http://play.integer32.com/?gist=4b6cc013ae2049c48445748d526fea2a&version=stable

(It runs ... it should output `Writing to 0002: 00 A9`).

**Comment by simon-whitehead** [comment]
Here is another one with the `AppSetting::NoBinaryName` setting applied - which means we don't have to supply a vector with "hakka" as the first item: http://play.integer32.com/?gist=1447d96f779330138a1e94062b034ae3&version=stable.

If you change this line:

    let input = vec!["memset", "02", "00", "A9"];

To:

    let input = vec!["--help"]; //"memset", "02", "00", "A9"];

It prints the help output.

**Comment by simon-whitehead** [comment]
(Sorry for being so chatty)

I'm also open to changing the commands and how they work if you have suggestions. I am aware the commands and the arguments don't really match the *nix or Windows styles. Happy to make them flag based, as in: `memset --location 02 --bytes 00 A9`. Your suggestions are welcome 😄

**Comment by seventh-chord** [comment]
From what I gather, clap is more geared towards command line arguments, rather than parsing actual commands. If we where to use it, I think it would make more sense to model each command as a separate clap `App`. In this case I think it would be simpler to just roll something custom.

Regarding the last message , I'm not actually very familiar with assembly, but the way the commands currently work, without flags, seems much more intuitive and involves less typing.

**Comment by simon-whitehead** [comment]
You make a good point. Although I'd say each command would be a `SubCommand`.

I am not opposed to rolling something custom - I had seen clap and thought it worth mentioning.

Are you interested in tackling this task? Your proposed solution with a `Command` struct and a `HashMap` for a lookup table sounds perfect.

**Comment by seventh-chord** [comment]
Ill give it a try, and post back once I have something working.

**Comment by seventh-chord** [comment]
After having a little conflict with the borrow checker I managed to throw together this prototype. I just use a `Vec` for now, as I was having some trouble with the `HashMap` (Each `Command` needs to be inserted in multiple locations, because it might have multiple names). 
~~I would also rather have used closures, as they could be typed out inside the `Command` declaration. Currently a function has to be declared for each command (Though the function can be declared inside another function)~~.
I would really appreciate it if you could give me some advice on improving it, as I am still rather new to rust :)

https://play.rust-lang.org/?gist=1840f7398f0947037c95de495e717692

Edit: Got closures working

**Comment by simon-whitehead** [comment]
A great start!

I worry it might involve quite a bit of manual lifetime pollution though, given the `&str` references being used everywhere.

The closure way is possible, but we will end up boxing everything and type signatures could get messy. I did try some generic closures but that too began to pollute everything with signatures like `where for <'b> F: Fn(&'b [&'b str])` .. which really started to look horrible :(

I had a play though and from a completely different angle, came up with this:

http://play.integer32.com/?gist=e2efc7ecf281059fd0dedcda714e2b43&version=stable

The idea basically is that we have each command be its own `struct` and we just dispatch to it based on the very first input argument. The dispatching takes care of wrapping up the arguments and providing them to the command in a wrapped type. How does this look to you? It removes the `&str` references and aims for dynamic dispatch rather than static. It will also remove the cyclic complexity of the current method because each `execute` will be in its own `struct`.

What do you think of this approach?

**Comment by seventh-chord** [comment]
I actually got the closures working with lifetime elision, but it still looked rather messy. Also, I didn't even think of that my approach would not resolve the complexity issue.

In general, your approach of having one `struct` per command seems a lot more organized. I do however not quite follow why you instanciate the struct whenever you want to run a command, in addition to having an instance of the struct stored in the command system.

I tried reworking this a bit. Command args are now passed to the `execute` method, and a single instance of the struct is used to execute the command. To run a command, a `Into<String>` is passed to the `CommandSystem`, which searches through the list of commands that have been added and executes one if the name matches. This also allows us to add commands from other places, as there no longer is a switch statement which needs to have entries for all commands.

I am a bit unsure about the `matches_name` function, maybe it should return a list of valid names, but just having an expression seems like the simplest way to do it.

~~http://play.integer32.com/?gist=e2efc7ecf281059fd0dedcda714e2b43&version=stable~~

http://play.integer32.com/?gist=492bc81f28b0b50ad8f88cd0c994f226&version=stable

**Comment by seventh-chord** [comment]
Another thing we need to figure out is output handling. Currently everything uses `println!`, but we will need to somehow pass a `&mut Console` to the command system so we can print to that. I have been playing around with implementing `std::io::Write` for the console, so we could use `writeln!` to print formatted text straight to the console. This means we could pass `&mut Write` instead.

**Comment by simon-whitehead** [comment]
@seventh-chord the `play.integer32.com` link you used in your comment is the same URL as mine. I think you might have accidentally re-pasted my link instead of yours? 😃 

Nevertheless, your description seems much more sane than mine. However I won't be able to comment on the `matches_name` function you've mentioned until you fix the playground link 😄

**Comment by simon-whitehead** [comment]
I got closures working if I boxed them, e.g.: `Box<Fn(...)>`, but I wasn't too sure about that approach. I had trouble getting the compiler to be happy eliding some of the lifetimes so I am keen to see how you managed that.

**Comment by seventh-chord** [comment]
Whoops. Correct link:

http://play.integer32.com/?gist=492bc81f28b0b50ad8f88cd0c994f226&version=stable

**Comment by seventh-chord** [comment]
Regarding the boxed closures, I had the following struct definition:
```rust
struct Command<'a> {
    names: Vec<&'a str>,
    help: &'a str,
    action: Box<Fn(&[&str], &CommandSystem)>
}
```
I just followed what the book said: https://doc.rust-lang.org/book/closures.html#returning-closures.

I was however not able to elide the lifetimes without using the `Box`, if I remember correctly

**Comment by simon-whitehead** [comment]
@seventh-chord I really like your adjustments. I think it looks very clean and really allows each command to grow independently.

I don't foresee any borrowing issues RE passing mutable references through this system. On the surface it looks like it will be pretty straight forward pass-throughs that shouldn't cause issues (fingers crossed? heh). Do you think we should head in this direction then? Or did you think the boxed closures were more ergonomic?

**Comment by seventh-chord** [comment]
@simon-whitehead Thank you :)

I think we should definitely go with the struct approach, considering how much cleaner they are then what I was initially suggesting.

**Comment by seventh-chord** [comment]
So I have started implementing this, and have hit a roadblock. As far as I know, it is not possible for `VirtualMachine`, which owns a `CommandSystem`, to borrow `&mut self` to the `CommandSystem`, but this is needed, as the `execute` function of the commands needs to modify the state of the `VirtualMachine`.

Edit: If I understand everything correctly, this means that the `CommandSystem` and `VirtualMachine` both need to be owned by some other function. I guess the `CommandSystem` should be located inside the `main.rs` of each level.

**Comment by simon-whitehead** [comment]
Heh - I guess my assumption that there wouldn't be any borrowing issues was completely wrong!

I wonder if the `CommandSystem` is and abstraction that we don't need. What if the `VirtualMachine` did all of the command setup and handling, and it worked somewhat like this:

    // ... a method on VirtualMachine
    pub fn handle_input(&mut self, input: &str) {
        let command = self.get_command(input);
        command.execute(self);
    }

This just takes the `CommandSystem` out of the equation and instead moves all its setup and logic into the `VirtualMachine`. Then, each command can be delegated to for the actual work.

Here is a small sample on the playground: http://play.integer32.com/?gist=676f01be1c1621d2802618751281b90d&version=stable

Again, I think this works _in theory_ (although I said that last time didn't I? heh). I also think the `CommandSystem` as you've implemented it currently is small enough in size that this is not too bad of a solution.

What do you think?

(By the way: Happy New Year! 😄 )

**Comment by seventh-chord** [comment]
Happy new year :tada: 

Well, the command system in itself is quite small, but with all commands added it will probably be quite a lot of code. Considering that each command is contained within its own struct I don't think readability will be an issue, so what you are suggesting sounds like a great solution. I'll give it a try!

Edit: Commands need mutable access to the virtual machine. You can never pass `&mut self` to a function, unless you are calling that function on `self` (That is, you own the function). Therefore, either all commands need to be functions in `impl VirtualMachine` or commands need to be executed outside the virtual machine. At least I understand the borrow checker better now :P

I have some working code with a `CommandSystem` struct. The disadvantage of this is that the struct has to be managed by the game (inside `fn main() { ... }`). We can move the command system to inside the vm (which would still require the game to process the commands), but I think what would be better is to have some struct containing `Console`, `VirtualMachine` and `CommandSystem`, which manages and updates all of those, and simply have the game create that struct.

**Comment by simon-whitehead** [comment]
I am happy with that approach as well I think (having another type that wraps all of it). I really want to avoid having the levels be the one managing the dependencies. Ideally, the level has `let everything_that_manages_the_vm_console_etc = Thing::new();`.. and thats all. If that requires another type to wrap it, I am happy with that.

> With all the commands added it would be quite a lot of code

I thought we would add another folder `vm/command_system`. Each `struct`/command could be its own file in there (allowing each to grow independently). For example:

    vm/
      |____ command_system
        |____ greet.rs
        |____ help.rs

 I imagine thats probably a good idea either way we go?

**Comment by seventh-chord** [comment]
I just finished implementing all commands. They currently are their own `impl` block in `command.rs`. I also looked into moving each command to its own file. While this helps readability, it adds some boilerplate with all the `use` and `mod` declarations, so I am unsure whether it is actually worth it. For reference, `command.rs` is now 290 lines. What do you think?

Also, I'll open a pull request, and would appreciate it if you could help me check if all the commands work as intended. I might have inadvertently introduced some bugs while rewriting the code.

**Comment by simon-whitehead** [comment]
This was merged in #23
